### PR TITLE
Specialisation for transfer along identical lattices

### PR DESCRIPTION
**Pull request by epolack**
Hope this is usefull (saw the `todo`).

### CODE REVIEW

File: src/interpolation.jl [hunk 1]
```diff
@@ -2,37 +2,64 @@ import Interpolations
...
-            supercell[i] = 1
-        else
-            supercell[i] = round(Int, norm(lattice_out[:, i]) / norm(lattice_in[:, i]))
-        end
-        if norm(lattice_out[:, i] - supercell[i]*lattice_in[:, i]) > .3*norm(lattice_out[:, i])
-            @warn "In direction $i, the output lattice is very different from the input lattice"
-        end
+    # Early exit if same lattice for input and output
+    lattice_in == lattice_out &&
+        return ρ_out = interpolate_density(ρ_in, grid_in, grid_out)
```
Comment (mfherbst) [inline src/interpolation.jl:43]: would write this as multi-line `if`.

File: src/interpolation.jl [hunk 2]
```diff
@@ -2,37 +2,64 @@ import Interpolations
...
+    @assert iszero.(eachcol(lattice_in)) == iszero.(eachcol(lattice_out))
+
+    # First, build supercell, array of 3 integers
+    supercell = map(zip(eachcol(lattice_in), eachcol(lattice_out))) do (col_in, col_out)
+        iszero(col_in) ? 1 : round(Int, norm(col_out) / norm(col_in))
+    end
+
+    # Check if some directions of lattice_in is not too big compared to lattice_out.
+    suspicious_directions = findall(norm.(eachcol(lattice_out - supercell .* lattice_in))
+                                    .> 0.3*norm.(eachcol(lattice_out)))
```
Comment (mfherbst) [inline src/interpolation.jl:55]: Use a  do ... end block

**Review by mfherbst — COMMENTED** [review]
Nice.

**Review by mfherbst — COMMENTED** [review]
Nice. Just some nits.

File: src/interpolation.jl [hunk 3]
```diff
@@ -11,23 +11,48 @@ function interpolate_density(ρ_in, basis_in::PlaneWaveBasis, basis_out::PlaneWa
...
+    ρ_out
+end
+
+function interpolate_density(ρ_in::AbstractArray, grid_in, grid_out, lattice_in,
+                             lattice_out=lattice_in)
     @assert size(ρ_in) == grid_in
 
+    # Early exit if same lattice for input and output
+    lattice_in == lattice_out &&
+        return ρ_out = interpolate_density_supercell(ρ_in, grid_in, grid_out)
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/interpolation.jl]: Yeah Antoine sometimes likes to do this for clarity.

[the remark this answers is not in the document]
Comment (antoine-levitt) [reply src/interpolation.jl]: Do I? I don't remember doing this on purpose :sweat_smile:

File: src/interpolation.jl [hunk 4]
```diff
@@ -51,11 +51,12 @@ function interpolate_density(ρ_in::AbstractArray, grid_in, grid_out, lattice_in
...
         # Check if some directions of lattice_in is not too big compared to lattice_out.
-        map(enumerate(eachcol(supercell .* lattice_in)),
-            eachcol(lattice_out)) do (i, scaled_col_in), col_out
-            norm(col_out - scaled_col_in) > 0.3 * norm(col_out) &&
-                @warn "In direction $i, the output lattice is very different from the " *
-                       "input lattice"
+        lattice_in_super = supercell .* lattice_in
+        suspicious_directions = findall(eachcol(lattice_in_super), eachcol(lattice_out)) do as_in, a_out
+            norm(as_in - a_out) > 0.3 norm(a_out)
+        end
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/interpolation.jl:57]: me neither as I just made a mistake ;)

**Pull request merged**